numpy/lib/function_base.py

```diff
@@ -3929,21 +3928,33 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        # the linear interpolation below ensure monotonicity with floating
+        # point operations and was proposed in:
+        # https://math.stackexchange.com/a/1798323 by Pedro Gimeno
+        diff_x2_x1 = x2 - x1
 
-        # ensure axis with q-th is first
         x1 = np.moveaxis(x1, axis, 0)
         x2 = np.moveaxis(x2, axis, 0)
 
+        r_above = np.add(x1, diff_x2_x1 * weights_above, out=out)
+        r_below = np.subtract(
+            x2, diff_x2_x1 * weights_below, out=r_above,
+            where=weights_above < 0.5
```

Comment:
This fun, but this way, please just call it `r` (or something longer) and not `above` and `below`, and add a comment that we first calculate everything for above, and then replace it with the "below" version.

Are there test for the combinations of "zerod" and `out=...` being given. Because this smells to me like we could get into trouble here.
Probably it is possible to move the `if zerod` block before these calculations to avoid that problem?

Reply:
I think the point is we made everything 1d at the top of the function, so the zerod block should remain below - to convert the final result back to 0d.

Perhaps we should push forward on the `leavewrapped` thing at some point.

Reply:
But `out` is never made 1-D, so we need to be careful?

Reply:
Yeah, this is a good example of for a need of `leavewrapped`.